src/JWT/JwtBuilder.cs

```diff
@@ -0,0 +1,266 @@
+        /// </summary>
+        /// <param name="provider">Zou custom provider you want use.</param>
+        /// <returns>The current builder-instance</returns>
+        public JwtBuilder SetTimeProvider(IDateTimeProvider provider)
+        {
+            _utcProvieder = provider;
+            return this;
+        }
+
+        /// <summary>
+        /// Set the validTor. This is requierd for decode with verification!
+        /// </summary>
+        /// <param name="validTor">Your custom validTor</param>
+        /// <returns>The current builder-instance</returns>        
+        public JwtBuilder SetValidTor(IJwtValidTor validTor)
+        {
+            _validTor = validTor;
+            return this;
+        }
+
+        /// <summary>
+        /// Set a custom URL encoder.
+        /// If you don't call this it will be use <see cref="JwtBase64UrlEncoder"/>
+        /// </summary>
+        /// <param name="urlEncoder">your custom encoder</param>
+        /// <returns>The current builder-instance</returns>
+        public JwtBuilder SetUrlEncoder(IBase64UrlEncoder urlEncoder)
+        {
+            _urlEncoder = urlEncoder;
+            return this;
+        }
+
+        /// <summary>
+        /// Set a algorithm. This is requierd if you want create a new Token!
+        /// </summary>
+        /// <param name="algorithm">your alogrithm to sign the JWT.</param>
+        /// <returns>The current builder-instance</returns>
+        public JwtBuilder SetAlgorithm(IJwtAlgorithm algorithm)
+        {
+            _algorithm = algorithm;
+            return this;
+        }
+
+        /// <summary>
+        /// Set a secret. This is requiered to build a new Token and to verify a token while decoding.
+        /// </summary>
+        /// <param name="secret">You secret to sign the token</param>
+        /// <returns>The current builder-instance</returns>
+        public JwtBuilder SetSecret(string secret)
+        {
+            _secret = secret;
+            return this;
+        }
+
+        /// <summary>
+        /// Tell the Decoder to check if the token is trusted!
+        /// </summary>
+        /// <returns>The current builder-instance</returns>
+        public JwtBuilder MustVerify()
+        {
+            _verify = true;
+            return this;
+        }
+
+        /// <summary>
+        /// Tell the Decoder to not check the token. This is default!
+        /// </summary>
+        /// <returns>The current builder-instance</returns>
+        public JwtBuilder NotVerify()
+        {
+            _verify = false;
+            return this;
+        }
+
+        /// <summary>
+        /// Build the token from the Information that this instance of <see cref="JwtBuilder"/> have.
+        /// </summary>
+        /// <returns>The JWT String.</returns>
+        public string Build()
+        {
+            if (!CanBuild())
+            {
+                throw new Exception("Can't build a Token. Check if you have call all of this: \n\r" +
```

Comment:
Do you think it should throw a more specific exception @abatishchev?

Reply:
It seems InvalidOperationException fits well here.